clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapChainStorage.java

```diff
@@ -0,0 +1,568 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ehcache.clustered.server.offheap;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import org.ehcache.clustered.common.store.Chain;
+import org.ehcache.clustered.common.store.Element;
+import org.ehcache.clustered.common.store.Util;
+import org.terracotta.offheapstore.paging.OffHeapStorageArea;
+import org.terracotta.offheapstore.paging.PageSource;
+import org.terracotta.offheapstore.storage.PointerSize;
+import org.terracotta.offheapstore.storage.StorageEngine;
+import org.terracotta.offheapstore.storage.portability.Portability;
+
+import static java.util.Collections.unmodifiableList;
+
+class OffHeapChainStorage<K> implements StorageEngine<K, InternalChain> {
+
+  private static final int ELEMENT_LENGTH = 0;
+  private static final int ELEMENT_NEXT = 4;
+  private static final int ELEMENT_HEADER_SIZE = 12;
+
+  private static final int CHAIN_KEY_LENGTH = 0;
+  private static final int CHAIN_KEY_HASH = 4;
+  private static final int CHAIN_TAIL = 8;
+  private static final int CHAIN_HEADER_SIZE = 16;
```

Comment:
As you probably guessed, they are all offsets in to the offheap data structures. I'll look in to rewording them to make them clearer/cleaner.